chapters/en/Unit 2 - Convolutional Neural Networks/mobilenet.mdx

```diff
@@ -0,0 +1,173 @@
+ 
+MobileNet also employs techniques like channel-wise linear bottleneck layers, which improve model accuracy while reducing the number of parameters. The architecture is designed with optimizations for various hardware platforms, including CPUs, GPUs, and even specialized hardware such as Google's Tensor Processing Units (TPUs).
+
+### Channel-wise Linear Bottleneck Layers
+Channel-wise linear bottleneck layers help to further reduce the number of parameters and computational cost while maintaining high accuracy in image classification tasks.
+
+A channel-wise linear bottleneck layer consists of three main operations applied sequentially:  
+1. **Depthwise convolution:** This step performs a convolution separately for each channel (a single color or feature) in the input image using a small filter (usually 3x3). The output of this step is the same size as the input but with fewer channels.  
+2. **Batch normalization:** This operation normalizes the activation values across each channel, helping to stabilize the training process and improve generalization performance.  
+3. **Activation function:** Typically, a ReLU (Rectified Linear Unit) activation function is used after batch normalization to introduce non-linearity in the network.
+
+### What does ReLU do?
+Some problems may occur during training. We will explain them first, then we explain what ReLU does to this problems. 
+
+#### Vanishing Gradient Problem:
+In deep neural networks, especially during backpropagation, the vanishing gradient problem can occur. This happens when gradients (which are used to update the network's weights) become very small as they are passed back through the network's layers. If these gradients become too small, they "vanish," making it hard for the network to learn and adjust its weights effectively.
+
+ReLU has a linear, non-saturating form for positive values (where it simply outputs the input if it's positive), it ensures that the gradients do not become too small, allowing for faster learning and more effective weight adjustments in the network.
+
+#### Non-Linearity:
+Without non-linearity, a neural network, no matter how many layers it has, would function as a linear model, incapable of learning complex patterns. 
+
+Non-linear functions like ReLU enable neural networks to capture and learn complex relationships in the data.
+
+You can find an example implementation of MobileNet for PyTorch below:
+
+PyTorch:
+```python  
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+class DepthwiseSeparableConv(nn.Module):
+    def __init__(self, in_channels, out_channels, stride):
+        super().__init__()
+        self.depthwise = nn.Conv2d(in_channels, in_channels, kernel_size=3, stride=stride, padding=1, groups=in_channels)
+        self.pointwise = nn.Conv2d(in_channels, out_channels, kernel_size=1, stride=1, padding=0)
+
```

Comment:
In the paper https://arxiv.org/pdf/1704.04861.pdf, they have included BN and ReLU layers as well. 
![image](https://github.com/sezan92/computer-vision-course/assets/34167018/3e2c0cc2-37be-41a6-8c5b-14b3a09b5d15)